docs/getting_started_with_jax_for_AI.md

```diff
@@ -0,0 +1,417 @@
+
++++ {"id": "1Y92oUSGeoRz"}
+
+## What does this tutorial cover?
+
+JAX itself focuses on array-based computation, and is at the core of a growing ecosystem of domain-specific tools. This tutorial introduces part of that ecosystem that is useful for defininig and training AI models, including:
+
+- [flax](http://flax.readthedocs.io): a tool designed for defining and building
+  scalable neural networks using JAX.
+- [optax](http://optax.readthedocs.io): a tool designed for high-performance
+  optimization of functions in JAX, including the loss functions used in neural network training.
+
+Once you've worked through this content, you may wish to visit http://jax.readthedocs.io/ for a deeper dive into the JAX library itself.
+
++++ {"id": "z7sAr0sderhh"}
+
+## Example: a simple neural network with flax
+
+We'll start with a very quick example of what it looks like to use JAX with the [flax](https://flax.readthedocs.io) framework to define and train a very simple neural network on a hand-written digits dataset.
+
++++ {"id": "pOlnhK-EioSk"}
+
+### Loading the data
+
+JAX can work with a variety of data loaders, but for simplicity here we can use the well-known [scikit-learn `digits`](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_digits.html) dataset.
+
+```{code-cell}
+:id: hKhPLnNxfOHU
+:outputId: ac3508f0-ccc6-409b-c719-99a4b8f94bd6
+
+from sklearn.datasets import load_digits
+digits = load_digits()
+
+print(f"{digits.data.shape=}")
+print(f"{digits.target.shape=}")
+```
+
++++ {"id": "lst3E34dgrLc"}
+
+This dataset consists of 8x8 pixelized images of hand-written digits along with labels, and we can visualize a handful of them this way:
+
+```{code-cell}
+:id: Y8cMntSdfyyT
+:outputId: 9343a558-cd8c-473c-c109-aa8015c7ae7e
+
+import matplotlib.pyplot as plt
+
+fig, axes = plt.subplots(10, 10, figsize=(6, 6),
+                         subplot_kw={'xticks':[], 'yticks':[]},
+                         gridspec_kw=dict(hspace=0.1, wspace=0.1))
+
+for i, ax in enumerate(axes.flat):
+    ax.imshow(digits.images[i], cmap='binary', interpolation='gaussian')
+    ax.text(0.05, 0.05, str(digits.target[i]), transform=ax.transAxes, color='green')
+```
+
++++ {"id": "Z3l45KgtfUUo"}
+
+Let's split these into a training and testing set, and convert these splits into JAX arrays which will be ready to feed into our model:
+
+```{code-cell}
+:id: 6jrYisoPh6TL
+
+from sklearn.model_selection import train_test_split
+splits = train_test_split(digits.images, digits.target, random_state=0)
+```
+
+```{code-cell}
+:id: oMRcwKd4hqOo
+:outputId: 0ad36290-397b-431d-eba2-ef114daf5ea6
+
+import jax.numpy as jnp
+images_train, images_test, label_train, label_test = map(jnp.asarray, splits)
+print(f"{images_train.shape=} {label_train.shape=}")
+print(f"{images_test.shape=}  {label_test.shape=}")
+```
+
++++ {"id": "JzrixENjifiq"}
+
+### Defining the flax model
```

Comment:
So this is the first time the user is seeing rngs.  Should we add a sentence about them why they are important?

Reply:
Good idea. Check this out:

```
**Note:** The [`flax.nnx.Module`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/module.html#flax.nnx.Module)
system is covered in more detail in the
[Flax basics](https://flax.readthedocs.io/en/latest/nnx_basics.html#the-flax-nnx-module-system) tutorial. Notice the use of the
[`flax.nnx.Rngs`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/rnglib.html#flax.nnx.Rngs) 
type - this is the main convenience API in Flax for managing random state.
The `Rngs` object can be can be used to get new unique keys based on a root JAX 
[pseudorandom number generator (PRNG)](https://jax.readthedocs.io/en/latest/random-numbers.html)
key passed to the constructor. Learn more about PRNGs in the Flax
[Randomness](https://flax.readthedocs.io/en/latest/guides/randomness.html)
and the JAX [PRNG](https://jax.readthedocs.io/en/latest/random-numbers.html)
tutorials.
```

cc @cgarciae

Reply:
https://github.com/jax-ml/jax-ai-stack/pull/54

Reply:
This is also the first time seeing several other things as well (e.g. `nnx.Module`, `nnx.Linear`, etc.) I'm not sure how to explain all of those things without distracting from the main flow. I think our users are smart: if they don't understand something here, they can google it.

Reply:
The context here: our initial attempt at JAX documentation tried to explain everything from the ground up, so that each time a user sees a concept it's fully explained: https://jax.readthedocs.io/en/latest/tutorials.html

The main feedback from that is that it's too boring and expository. So we're creating this tutorial flow as a response. For that reason, we're *explicitly* choosing not to explain every concept as it comes up, because we already have a doc flow that does that, and people don't like it.